Pre-work from the forex terminology PR

tcharding opened the pull request:
This PR is some the initial pre-work patches from https://github.com/comit-network/comit-rs/pull/2983

This stuff is useful on its own even though the motivation comes from the forex work.

Please review carefully the functions added/modified in the `ethereum` module.

D4nte commented on comit/src/asset/ethereum/erc20.rs:
```diff
@@ -22,11 +27,18 @@ impl Erc20Quantity {
 
+    /// Create a erc20 quantity from a float string. E.g., this can be used to
+    /// create a DAI quantity from the ergonomic string "1234.56".
+    pub fn try_from_float(eth: &str) -> anyhow::Result<Self> {
```
```suggestion
    pub fn try_from_eth_str(eth: &str) -> anyhow::Result<Self> {
```
Let's try to keep the function names similar.

tcharding replied:
Well picked up (the inconsistency), thanks. Will keep in mind (the ethereum stuff is on hold until float_math goes in).

D4nte commented on comit/src/asset/ethereum/erc20.rs:
```diff
@@ -22,11 +27,18 @@ impl Erc20Quantity {
     }
 
-    pub fn from_wei_dec_str(str: &str) -> Result<Self, Error> {
+    pub fn try_from_wei_dec_str(str: &str) -> Result<Self, Error> {
```
```suggestion
    pub fn try_from_wei_dec_str(wei: &str) -> Result<Self, Error> {
```
Just to make it clearer as the type is already expressed in the signature.

tcharding replied:
Better, thanks.

D4nte commented on comit/src/float_math.rs:
```diff
@@ -0,0 +1,223 @@
+use bitcoin::hashes::core::cmp::Ordering;
```
So actually the name of the file is incorrect 😁 We are doing fixed point arithmetic here.
The aim of the module is to brought some safety in handling fractional numbers and doing maths/arithmetic with them. Some ideas for a better name:
- fraction
- fractional number
- maths
- arithmetic
- fixed point

tcharding replied:
+1 for `fraction`

D4nte commented on comit/src/float_math.rs:
```diff
@@ -0,0 +1,223 @@
+}
+
+/// Truncate the float's mantissa to length `precision`.
+pub fn truncate_str(float: &str, precision: u16) -> f64 {
```
I think it would make more sense for it to return a string as it took a string slice as input.  the conversion back in a float can be done in `truncate`.

```suggestion
pub fn truncate_str(number: &str, precision: u16) -> f64 {
```

tcharding replied:
Aggree.

D4nte commented on comit/src/float_math.rs:
```diff
@@ -0,0 +1,223 @@
+    truncate_str(&s, precision)
+}
+
+/// Truncate the float's mantissa to length `precision`.
```
```suggestion
/// Truncate the number's mantissa to length `precision`.
```

I have loosely used the word `float` when I actually mean `number` or more precisely `fractional number`.

tcharding replied:
How about we favour the term `fraction` in identifiers if we choose to use that as the filename?

D4nte commented on comit/src/float_math.rs:
```diff
@@ -0,0 +1,223 @@
+
+/// Truncate the float's mantissa to length `precision`.
+pub fn truncate_str(float: &str, precision: u16) -> f64 {
+    let mut string = float.to_string();
```
Because this accepts a string, there is no guarantee the string actually represents a number.
We should check that and also return a result.
Also, we should proptest this function with any arbitrary string value and ensure it does not panic/overflows.

tcharding replied:
Oh, you mean we should actually write the API properly instead of me just hacking ups some shit and trying to merge it into the comit lib - if this is what you mean you are totally correct. Sloppy work here from me. I think we should move comit to a separate crate and enforce much stricter code practices on it.

D4nte commented on comit/src/float_math.rs:
```diff
@@ -0,0 +1,223 @@
+    let index = string.find('.');
+
+    match index {
+        None => f64::from_str(float).expect("This has no decimal, should convert"),
```
This would panic if the string passed is not a number.

D4nte commented on comit/src/float_math.rs:
```diff
@@ -0,0 +1,223 @@
+    match decimal_index {
+        None => {
+            let zeroes = "0".repeat(pow as usize);
+            Ok(BigUint::from_str(&format!("{}{}", float, zeroes)).expect("an integer"))
```
It accepts any string so that can lead to a panic. Validate the string and proptest looking for panics.

D4nte commented on comit/src/network/protocols/orderbook.rs:
```diff
@@ -29,8 +30,10 @@ use std::{
-const BTC_DAI: &str = "BTC/DAI";
+lazy_static::lazy_static! {
+    /// The Topic used to publish BTC/DAI orders.
+    static ref BTC_DAI_TOPIC: Topic = Topic::new("BTC/DAI".to_string());
```
So, I guess I should start a mailing list discussion about the topic?

tcharding replied:
As I've said in other places, the whole gossipsub needs rethinking/redesigning IMNSHO. Please do hit the mailing list, I'm interested in you view, I have some ideas also.

tcharding commented:
@D4nte, I did not go over yet your ideas re foat_math module. I have removed it from this PR simply because we shouldn't push anything into COMIT lib that is not spot on. This PR is meant to be simple pre-work, probably should never have included the float_math stuff. I also split it into two separate PRs , one for ethereum stuff and one for orderbook stuff. I'd like to flag this PR somehow so your suggestions are not lost, I've saved it as a bookmark - any better workflow ideas welcome.

tcharding commented:
Kicking in the stripped down PR with just the approved patches: bors r+

tcharding commented:
bors r+

tcharding commented:
bors r+

tcharding commented:
bors r+

tcharding commented:
bors r+

tcharding commented:
bors r+
